### PR TITLE
feat: support cache mounts

### PR DESCRIPTION
**Pull request by frezbo**
Support cache mounts

Signed-off-by: Noel Georgi <git@frezbo.dev>

### CODE REVIEW

**Comment by frezbo** [comment]
Tested using a sample `pkg.yaml`

```yaml
name: test
variant: scratch
shell: /toolchain/bin/bash
dependencies:
  - stage: base
steps:
  - cache: /.cache
    prepare:
      - |
        mkdir -p /usr/bin \
          && ln -sf /toolchain/bin/env /usr/bin/env \
    build:
      - |
        mkdir -p /rootfs
        
        ccache-swip --help > /rootfs/ccache || true
        if [[ -f /.cache/file ]]; then
          val=$(cat /.cache/file)
          val=$((val+1))
          echo $val > /.cache/file
          echo $val > /rootfs/value
        else
          echo 1 > /.cache/file
        fi
  # {{ else }}
  - install:
      - |
        mkdir -p /rootfs
# {{ end }}
finalize:
  - from: /rootfs
    to: /
```

manually confirmed that the `value` is incremented with every build that invalidates the cache

File: internal/pkg/convert/node.go [hunk 1]
```diff
@@ -214,16 +215,26 @@ func (node *NodeLLB) stepScripts(root llb.State, i int, step v1alpha2.Step) llb.
...
-					llb.Args([]string{
-						node.Pkg.Shell.Get(),
-						"-c",
-						instruction.Script(),
-					}),
-					llb.WithCustomName(fmt.Sprintf("%s%s-%d", node.Prefix, script.Desc, i)),
-				)...,
-			).Root()
+			mountOpts := make([]llb.MountOption, 0, 1)
+			runOptions := node.Graph.commonRunOptions
```
Comment (smira) [inline internal/pkg/convert/node.go:219]:
```suggestion
			runOptions := append([]llb.RunOptions(nil), node.Graph.commonRunOptions...)
```

As `append` might either modify existing slice or re-allocate to new one, we need to protect original value from being changed.

File: internal/pkg/convert/node.go [hunk 2]
```diff
@@ -214,16 +215,26 @@ func (node *NodeLLB) stepScripts(root llb.State, i int, step v1alpha2.Step) llb.
...
-				append(node.Graph.commonRunOptions,
-					llb.Args([]string{
-						node.Pkg.Shell.Get(),
-						"-c",
-						instruction.Script(),
-					}),
-					llb.WithCustomName(fmt.Sprintf("%s%s-%d", node.Prefix, script.Desc, i)),
-				)...,
-			).Root()
+			mountOpts := make([]llb.MountOption, 0, 1)
```
Comment (smira) [inline internal/pkg/convert/node.go:218]: we can move this into `if` statement

File: internal/pkg/convert/node.go [hunk 3]
```diff
@@ -214,16 +215,26 @@ func (node *NodeLLB) stepScripts(root llb.State, i int, step v1alpha2.Step) llb.
...
+			runOptions := node.Graph.commonRunOptions
+
+			if step.CachePath != "" {
+				sharing := llb.CacheMountShared
+
+				mountOpts = append(mountOpts, llb.AsPersistentCacheDir(path.Clean(step.CachePath), sharing))
+				runOptions = append(runOptions, llb.AddMount(step.CachePath, root, mountOpts...))
```
Comment (smira) [inline internal/pkg/convert/node.go:225]:
```suggestion
				runOptions = append(runOptions, llb.AddMount(step.CachePath, root, llb.AsPersistentCacheDir(path.Clean(step.CachePath), llb.CacheMountShared)))
```

**Review by smira — APPROVED** [review]
it might be also nice to update `README.md` with the new option

**Pull request merged**